ql/src/java/org/apache/hadoop/hive/ql/parse/QB.java

```diff
@@ -456,4 +456,22 @@ public boolean containsQueryWithoutSourceTable() {
     }
     return aliasToTabs.size()==0 && aliasToSubq.size()==0;
   }
+
+  // returns false when the query block doesn't have
+  // a table defined, e.g. "select 5"
+  public boolean hasTableDefined() {
+    // A table is not defined when there only exists a dummy
+    // table in the alias, so if it is blank, we know a table
+    // is defined.
+    if (aliasToTabs.size() == 0) {
```

Comment:
This is a bit counterintuitive... Just to confirm, if `.size() == 0`, then there are table definitions in the query?

Reply:
This seemed weird to me too, and perhaps I need a better solution.  I may be mistaken here, but I had tests fail when the table had no alias (e.g. 'select * from tbl1' as opposed to 'select * from tbl1 alias1').  However, when we had no table, there WAS an alias, and it was "DUMMY_TABLE";

I didn't see any structure within QB containing all table names.  Is there another place you think I can grab it?

Reply:
Can we check when you were getting `null`? It is suspicious... AFAIK, before calling this method, we iterate through the same structure in L14102 in SemanticAnalyzer and alias should not be null (btw, this code could be integrated within this loop to avoid looping twice).

Reply:
Actually, it's more than suspicious...I have a bug (was just thinking about it)

Even if it is true that the size is 0 when there are no tables, there will be a bug when there is one table that has one subquery that has no table.  

So here's my proposal...put in an if check that returns false if there is exactly one table defined, it is a DUMMY_TABLE, and it doesn't have any subqueries (it looks like QB has that information).  Does that sound good?